linera-chain/src/data_types.rs

```diff
@@ -587,16 +587,62 @@ impl<'a> SignatureAggregator<'a> {
     }
 }
 
+fn is_ordered(values: &[(ValidatorName, Signature)]) -> bool {
+    values.windows(2).all(|pair| pair[0].0 <= pair[1].0)
+}
+
+impl<'de> Deserialize<'de> for Certificate {
+    fn deserialize<D>(deserializer: D) -> Result<Self, D::Error>
+    where
+        D: Deserializer<'de>,
+    {
+        struct CertificateVisitor;
+
+        impl<'de> serde::de::Visitor<'de> for CertificateVisitor {
+            type Value = Certificate;
+
+            fn expecting(&self, formatter: &mut std::fmt::Formatter) -> std::fmt::Result {
+                formatter.write_str("struct Certificate")
+            }
+
+            fn visit_seq<A>(self, mut seq: A) -> Result<Self::Value, A::Error>
+            where
+                A: serde::de::SeqAccess<'de>,
+            {
+                let value: HashedValue = seq
+                    .next_element()?
+                    .ok_or_else(|| serde::de::Error::invalid_length(0, &self))?;
+
+                let round: RoundNumber = seq
+                    .next_element()?
+                    .ok_or_else(|| serde::de::Error::invalid_length(1, &self))?;
+
+                let mut signatures: Vec<(ValidatorName, Signature)> = seq
+                    .next_element()?
+                    .ok_or_else(|| serde::de::Error::invalid_length(2, &self))?;
+
+                Ok(Certificate::new(value, round, &mut signatures))
```

Comment:
I'd rather fail and return an error if they aren't already sorted. That way each certificate has only one serialized form.

I _do_ think it's fine for `new` to do the sorting; I just wouldn't do it when deserializing.